### DISCUSSION TITLE
Additional ControlNet models

### DISCUSSION THREAD

**Discussion by agsandrew** [lllyasviel/ControlNet#114] [Ideas]
Please consider training the following models:

- Threshold model. Basic 50% threshold can provide essential guidance for image recreation such as separation of background and foreground, areas of illumination and shadows, essential edges, etc. Addition of noise to the original image can reveal halftone details if necessary.

- Classic Photoshop Find Edges or Glowing Edges model. Photoshop output of these two filters already works well enough as pre-processor with HUE model, but dedicated model can provide better results. Glowing Edges filter in particular provides fine control of image relevant edge details and some form of these filters exist practically in every image editing app out there to allow users great degree of flexibility.

- Helmholtz Analysis model. One can see output of this analysis in action in ImageJ (Main Menu -> Analyze -> Helmholtz Analysis). In essence its output allows to identify directionality of graphic elements and can be very useful in guiding recreation of fine flowing textures/lines such as hairs, paint strokes, organic patterns, etc.

- Line Skeleton model. Similar to Canny Edge but identifies center-lines of filled areas instead of their edges. One can see this in action in ImageJ (Main Menu -> Process -> Binary -> Skeletonize). This would allow additional control in recreation of images based on lines such as ink drawings, pencil sketches, manga, etc.

Thank you for great tool!

**Comment by geroldmeisinger** [comment]
Great ideas btw!

> Threshold model

If by threshold you mean luminance, I think I created something like this already here: https://civitai.com/models/137638?modelVersionId=151969 (click on "show more" and scroll down to "Training"). Conditioning images are converted to white-gray-black images. As for controlling luminance and brightness itself, look here https://civitai.com/models/80536/lighting-based-picture-control-controlnet

I don't understand how you would separate background and foreground however? I think depth would be a better fit for this or implementing Midas in image editors (TODO: has this been done actually?)

> Find Edges or Glowing Edges model

What would be really helpful in supporting your argument would be an image examples study:
* Pick carefully selected topics (landscape, person, architecture etc. or just 10 portraits if it works particular well for portraits)
* Apply classic canny edge detection
* Apply your "find edges" and "glowing edges" algorithm
* Evaluate and annotate areas where "find edges" or "glowing edges" produces better results ("look, it better outlines the face details in most cases, it even catches wrinkles and birthmarks")
* Even better: discuss how different parameters produce better or worse results so we know which parameter-spread we should use for training ("best results are usually between px=0.5-2.5, below breaks edges too much, above is basically just noise")

![image](https://github.com/lllyasviel/ControlNet/assets/112266044/ece73aff-0c4c-4872-89d4-80763ec79e84)

We have many edge detection algorithms, even deep learning based ones. Training isn't much work actually and we might even say "train'em all" but I'd still question "what's the point?". Having said that, I would be willing to give it a try and train a prototype if the approach looks promising for further evaluation. (I'm actually training https://github.com/lllyasviel/ControlNet/discussions/318 right now)

> Helmholtz Analysis

This sound VERY interesting. Could you please provide a few example images. If you find a python or imagemagick implementation, that would be very helpful. I'd be willing to give it a try.

> Line Skeleton

The same arguments as for "find edges" apply here. Could you elaborate more on how you think this would provide additional guidance as compared to, say, line drawings alone, which already provide enough "outline/shape/form"? Think about it like this: when we already have a original image (e.g. a photograph) we can derive depth or canny map and use it as a reference for guidance. if we do not have a reference we need draw something ourself (e.g. scribble, or generate the pose in a pose editor first). In which way would "skeleton" help here? It can be easily derived, but would only guide "extrusions" which is hard to do right I guess. for "outline/shape/form" we already have canny edges. would it capture a different level of detail which is useful (sometimes less is actually more)? Could "skeleton" help to come up with something ourself, like is there a drawing-method were you would start with skeleton-first (compare with Scribble)?

I discuss further idea generations in my article here https://civitai.com/articles/2078#heading-3521
